scripts/Tools/case.setup

```diff
@@ -1,475 +1,316 @@
+
+###############################################################################
+def _build_usernl_files(case, model, comp):
+###############################################################################
+    """
+    Create user_nl_xxx files, expects cwd is caseroot
+    """
+    model = model.upper()
+    model_file = case.get_value("CONFIG_%s_FILE" % model)
+    model_dir = os.path.dirname(model_file)
+
+    expect(os.path.isdir(model_dir),
+           "cannot find cime_config directory %s for component %s" % (model_dir, comp))
+
+    if comp == "cpl":
+        if not os.path.exists("user_nl_cpl"):
+            shutil.copy(os.path.join(model_dir, "user_nl_cpl"), ".")
+    else:
+        ninst = case.get_value("NINST_%s" % model)
+        nlfile = "user_nl_%s" % comp
+        model_nl = os.path.join(model_dir, nlfile)
+        if os.path.exists(model_nl):
+            if ninst > 1:
+                for inst_counter in xrange(1, ninst+1):
+                    case_nlfile = "%s_%04d" % (nlfile, inst_counter)
+                    if not os.path.exists(case_nlfile):
+                        shutil.copy(model_nl, case_nlfile)
+            else:
+                if not os.path.exists(nlfile):
+                    shutil.copy(model_nl, nlfile)
+
+###############################################################################
+def case_setup(caseroot, clean, test_mode):
+###############################################################################
+    os.chdir(caseroot)
+
+    cimeroot = os.environ["CIMEROOT"]
+
+    case = Case()
+
+    # Check that $DIN_LOC_ROOT exists - and abort if not a namelist compare tests
+    din_loc_root = case.get_value("DIN_LOC_ROOT")
+    testcase     = case.get_value("TESTCASE")
+    expect(not (not os.path.isdir(din_loc_root) and testcase != "SBN"),
+           "inputdata root is not a directory: \"$din_loc_root\" ")
+
+    # Check that userdefine settings are specified before expanding variable
+    for vid, value in case:
+        expect(not (type(value) is str and "USERDEFINED_required_build" in value),
+               "Parameter '%s' must be defined" % vid)
+
+    # Create batch script
+    if not clean:
+        mach = case.get_value("MACH")
+        expect(mach is not None, "xml variable MACH is not set")
+
+        # Create Macros file only if it does not exist
+        if not os.path.exists("Macros"):
+            logger.debug("Creating Macros file for %s" % mach)
+            set_compiler(case=case)
+        else:
+            logger.debug("Macros script already created ...skipping")
+
+        # Set tasks to 1 if mpi-serial library
+        if case.get_value("MPILIB") == "mpi-serial":
+            for vid, value in case:
+                if vid.startswith("NTASKS_") and value != 1:
+                    case.set_value(vid, 1)
+
+        # Check ninst. JGF: What does NINST stand for? Less hardcoded way to get comps?
+        for comp in ["ATM", "LND", "ROF", "OCN", "ICE", "GLC", "WAV"]:
```

Comment:
The prefered method to get comps is 
        files = Files()
        drv_config_file = files.get_value("CONFIG_DRV_FILE")
        drv_comp = Component(drv_config_file)
        component_classes = drv_comp.get_valid_model_components()